### ISSUE TITLE
Internal python errors printed when the scheduler has too many jobs to process

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#336]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/336)
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 4, 2018, 00:25

In larger projects, running `bst build` can result in these errors being printed:

```
BlockingIOError: [Errno 11] Resource temporarily unavailable:
Exception ignored when trying to write to the signal wakeup fd:
```

- https://gitlab.gnome.org/GNOME/gnome-build-meta/-/jobs/21193
- https://gitlab.gnome.org/GNOME/gnome-build-meta/-/jobs/21194

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 4, 2018, 05:40

It looks like we're hitting: https://bugs.python.org/issue21595

The above issue was *improved* since before apparently, but still not rock solid.

Note from the end of that thread:

> With the patch, I start getting messages with 140 concurrent processes, which is much better :-) IMO more than 100 concurrent processes is crazy, don't do that at home :-) I mean processes with a very short lifetime. The limit is the number of SIGCHLD per second, so the number of processes which end at the same second.

It seems quite likely to me that we would have more than 140 processes end at the same second, especially processes which found that there was nothing to do and exited immediately.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 4, 2018, 05:42

changed title from **{-Error when running bst --track in big projects: "BlockingIOError: [Errno 11] Resource temporarily unavailable: Exception ignored when trying to write to the signal wakeup fd:"-}** to **{+Internal python errors printed when running bst build --track on large projects+}**

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 4, 2018, 05:42

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on May 23, 2019, 08:22

mentioned in merge request !1344

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Dec 3, 2019, 10:40

This is still occurring for me heavily on the synthetic debian target, with a linux host https://gitlab.com/BuildStream/benchmarking/debian-stretch-bst/tree/jennis/add_all_files

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Dec 3, 2019, 10:56

Just wanted to check if this issue is for bst 1.x branch? (Since `bst build --track` doesn't exist on master anymore.)

If so, we should probably clarify that in the issue description somewhere.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 3, 2019, 11:09

This still happen for any operation that involves running jobs :(

It also happens on real projects (Could reproduce on Freedesktop-sdk)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Dec 3, 2019, 12:41

> This still happen for any operation that involves running jobs :(

Ah! In that case, can one of you please update the issue title/description to reflect that?

I'd do that myself but I don't have all the details.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Dec 4, 2019, 09:06

Another upstream ticket related to the error message https://bugs.python.org/issue32776

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Dec 5, 2019, 16:43

mentioned in merge request !1745

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 6, 2019, 08:04

Based on a quick test we can monkey patch to work around this issue:
```python
__real_set_wakeup_fd = signal.set_wakeup_fd

def __set_wakeup_fd_without_warning(fd, *, warn_on_full_buffer=True):
    __real_set_wakeup_fd(fd, warn_on_full_buffer=False)

signal.set_wakeup_fd = __set_wakeup_fd_without_warning
```

Not nice that we have to do this but it's simple enough and doesn't rely on any internal API.

As I understand it, this is safe with regards to `SIGCHLD` handling. I.e., for `SIGCHLD` the only purpose of the write to the socket is to wake up the event loop such that the child watcher can call `waitpid()`. It doesn't matter at all how many times `SIGCHLD` was raised between invocations of the child watcher.

We may lose other signals, though. E.g., if the socket buffer is full because of frequent `SIGCHLD` signals, `SIGINT` or `SIGTERM` will get lost as well and the corresponding warning would also be suppressed. However, these signals would be lost with or without the monkey patching and thus, I think we should do this.

A better solution might require replacing at least part of the event loop implementation. Signal handlers could set a flag (or increase a counter) instead of relying on the contents of the socket buffer.

Addendum: `warn_on_full_buffer` requires Python 3.7, however, we can easily make this conditional on the Python version.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 6, 2019, 09:31

I believe we can do better, though it might require a bit more work:

The reason why we get those messages, is because we are not able of processing enough messages per second.

This comes from two parts I can see:

The first part is [SafeChildWatcher._do_waitpid_all](https://github.com/python/cpython/blob/master/Lib/asyncio/unix_events.py#L1049) which scales linearly with the number of children it has to watch, and may awaken _multiple_ jobs completion, rendering next calls moot, as they won't find anything.

The second part is that the callback we have is not part of the async loop, it is something that is synchronous and ends up calling [Scheduler._sched](https://gitlab.com/BuildStream/buildstream/blob/master/src/buildstream/_scheduler/scheduler.py#L318), which will look for 1 _or more_ jobs to start.

What I propose is:

1) Re-implement part of the `SafeChildWatcher`, to count the number of awaken children and skip entirely the next `N` `SIGCHLD` messages. This would reduce the amount of work immediately done.

2) Transform the callback for job completion and _everything it touches_ to be asynchronous tasks. This will ensure that we can get interrupted in the _middle_ of the processing to handle more messages if needed.

3) Since we would now have `_sched` as a future, only schedule a new one if one is not _already_ scheduled.


I believe that with those three changes, we would end up not seeing this message anymore, or at least not until we get _way more_ concurrent jobs. It should also result in speeding up the whole scheduler.

If we can wait on this to be implemented and are happy with the proposal, I'm ready to implement it, and believe I can get it done before EOY.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 6, 2019, 09:32

changed title from **Internal python errors printed when {-running bst build --track on large project-}s** to **Internal python errors printed when {+the scheduler has too many jobs to proces+}s**

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 6, 2019, 09:32

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 6, 2019, 10:12

> Re-implement part of the `SafeChildWatcher`, to count the number of awaken children and skip entirely the next `N` `SIGCHLD` messages. This would reduce the amount of work immediately done.

I think this would risk a deadlock. As I understand it, there is no guarantee that the `SIGCHLD` handler is called exactly once per terminated child even in regular POSIX/C, e.g., if multiple children terminate while the signal is blocked (be it explicitly blocked or because the signal handler takes a while to be scheduled or complete). And with Python `set_wakeup_fd` there is an additional risk of receiving fewer `SIGCHLD`.

An alternative could be for the replacement `SafeChildWatcher` to install a Python signal handler for `SIGCHLD` in addition to the asyncio signal handler. That signal handler would simply set a global variable `sigchld_pending` or similar to `True`. At the beginning of the watcher's main handler, i.e., before the `waitpid` calls, immediately return if `sigchld_pending` is `False`. If it's `True`, set it to `False` and then perform the `waitpid` calls. If I got this right in my head, this would be a race free optimization without deadlock risk.

I'd prefer fixing the event loop such that it doesn't care about the contents of the socket buffer. I.e., use an approach similar to the one described in the previous paragraph but for all event loop signal handlers. Standard UNIX signals are not queued, so let's also skip the queueing internally. I wouldn't expect this to be more work overall, assuming we can replace the `add_signal_handler()` and co. of the standard event loop as I think we can.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 6, 2019, 10:20

> An alternative could be for the replacement `SafeChildWatcher` to install a Python signal handler for `SIGCHLD` in addition to the asyncio signal handler. That signal handler would simply set a global variable `sigchld_pending` or similar to `True`. At the beginning of the watcher's main handler, i.e., before the `waitpid` calls, immediately return if `sigchld_pending` is `False`. If it's `True`, set it to `False` and then perform the `waitpid` calls. If I got this right in my head, this would be a race free optimization without deadlock risk.

I like that, it should indeed work and be safer. (and nicer). I'll put up a MR for the SIGCHLD. Other signals should not happen very often so I'm less concerned about them.

Do we want also to go for the other optimizations describe here?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 6, 2019, 10:42

> Other signals should not happen very often so I'm less concerned about them.

It's true that there is virtually zero risk of the socket buffer being filled with lots of other signals. However, if the socket buffer gets filled with `SIGCHLD` despite our optimizations, we would still lose `SIGINT`, i.e., Ctrl+C would be non-functional for the user in that scenario.

That's why I'd prefer fixing it closer to the root cause, which is the event loop class, not the child watcher class. That way the `pending` flag for `SIGINT` and co. would reliably get set in the signal handler even if the socket buffer is already full. And we could safely pass `warn_on_full_buffer=True` (on Python 3.7+). Do you think fixing it in a custom event loop would be a lot more work than working around the issue in a custom child watcher?

> Do we want also to go for the other optimizations describe here?

Do you think they would provide a significant reduction in actual CPU time or is your only goal to avoid the signal issue? Keep in mind that invoking the O(n) child signal handler more frequently (because everything is async) might also sometimes increase actual CPU time.

I'm also not sure whether the async approach will make the code harder to read/maintain or not. If the only goal is to fix the signal handling issue, the suggested event loop fix alone should already be a complete and proper solution. That said, on Python 3.5/3.6 we would still not be able to suppress the warning even though a full socket/pipe buffer could be safely ignored. However, I don't think we should switch to an async approach for the sole reason of this warning on old Python (and we could probably mitigate the issue by creating a pipe with a much larger buffer on Python 3.5/3.6).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 6, 2019, 10:49

> It's true that there is virtually zero risk of the socket buffer being filled with lots of other signals. However, if the socket buffer gets filled with `SIGCHLD` despite our optimizations, we would still lose `SIGINT`, i.e., Ctrl+C would be non-functional for the user in that scenario.

True, however if we respect asyncio's standards of fast methods, we would not end up with the problem at all, since they would be able to be processed in time

> Do you think they would provide a significant reduction in actual CPU time or is your only goal to avoid the signal issue? Keep in mind that invoking the O(n) child signal handler more frequently (because everything is async) might also sometimes increase actual CPU time.

True, however, having this method setting wether we received a SIGCHLD since last time or not would probably be enough to avoid that being called too often. Note that we already call it once per SIGCHLD message received.

> the suggested event loop fix alone should already be a complete and proper solution.

I'm not convinced about that. That would tie very closely the child watcher's implementation with the event loop, making it harder to potentially later split that (for example when we want to support Windows)

I believe at least 

> Since we would now have `_sched` as a future, only schedule a new one if one is not *already* scheduled.

Should be implemented, as this is a real performance gain, as the _sched method is currently quite intensive

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 6, 2019, 10:54

mentioned in merge request !1755

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 6, 2019, 10:55

[[Gitlab user @juergbi]](https://gitlab.com/juergbi) for point 3, I would be thinking about something akin to: https://gitlab.com/BuildStream/buildstream/merge_requests/1755

I'll obviously need to benchmark this extensively still :)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 6, 2019, 11:06

> > the suggested event loop fix alone should already be a complete and proper solution.
> 
> I'm not convinced about that. That would tie very closely the child watcher's implementation with the event loop, making it harder to potentially later split that (for example when we want to support Windows)

No, the event loop fix as I imagine it would not be tied to a child watcher implementation. It would fix signal handling for all code that uses asyncio signal handling including (unmodified) `SafeChildWatcher`. At least when we're talking about standard UNIX signals, not real-time signals. It might even be possible to submit a cpython patch upstream based on this.

That said, it might only work for the UNIX `SelectorEventLoop` as I know nothing about the Windows `ProactorEventLoop`. However, in that case we would simply use the Python default event loop on other platforms. It wouldn't make it worse for those. However, it's a non-trivial topic and I could be missing something, of course.

> > Since we would now have `_sched` as a future, only schedule a new one if one is not *already* scheduled.
>
> Should be implemented, as this is a real performance gain, as the _sched method is currently quite intensive

I agree, something like !1755 makes sense to me.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 6, 2019, 11:16

That makes sense, I'll implement our own SelectorEventLoop to optimize this part of the handling and open a MR with it.

> I agree, something like !1755 makes sense to me.

Perfect, I'll benchmark this then remove the WIP status

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 9, 2019, 10:02

I believe !1755 is enough to remove the problem entirely. I've been testing it with bigger pipelines (32+ fetchers, 32+ builders) and was not able to reproduce the problem.

I'd suggest therefore closing this and reopening if/when we hit this bug again, at which point we should optimize the event loop. I believe wanting to optimize the event loop now is going too deep inside the asyncio internals (we would need to override private methods) and we already have our shares of problems around the event loop :)

Would that seem sensible to you? Would someone be happy to validate my findings?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 9, 2019, 10:18

I can also no longer reproduce the issue. Let's close it. We can always reopen if it starts happening again despite the optimizations.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 9, 2019, 10:18

closed
